File: src/scan.py
from more_itertools import peekable
from .utils import *

#
# Indentation preprocessor 
#

def without_comments(token_stream):
    for token in token_stream:
        if token.tag != TokenTag.com:
            yield token

def indent(token_stream):
    stream = peekable(without_comments(token_stream))
    levels = []

    def pop_levels(count=None):
        for _ in range(count or len(levels)):
            tok = levels.pop()
            yield mktoken(tag=TokenTag.ded, value='$DED', start=tok.start, stop=tok.stop)
        if stream:
            tok = stream[0]
            yield mktoken(tag=TokenTag.cnt, value='$CNT', start=tok.start, stop=tok.stop)

    for t in stream:
        if not is_whitespace(t):
            yield t

        elif t.tag == TokenTag.nl and stream:
            peek = next(stream)
            next_level = peek.value
            last_level = levels[-1].value if levels else ''

            if peek.tag != TokenTag.ws:
                for tok in pop_levels(): yield tok
                yield peek

            elif stream and stream[0].tag == TokenTag.nl: 
                continue

            elif next_level == last_level:
                yield mktoken(tag=TokenTag.cnt, value='$CNT', start=peek.start, stop=peek.stop)

            elif next_level.startswith(last_level):
                peek.diff = next_level[len(last_level):]
                levels.append(peek)
                yield mktoken(tag=TokenTag.ind, value='$IND', start=peek.start, stop=peek.stop)

            elif last_level.startswith(next_level):
                level = index(lambda x: x.value == next_level, reversed(levels))
                if level is None: raise indent_error(levels, peek)
                for tok in pop_levels(level): yield tok

            else:
                raise indent_error(levels, peek)

    for tok in pop_levels(): yield tok

#
# Basic table scanner
#

line, char = 1, 1
stream = None
delims = None
parens = {')': '(', ']': '[', '}': '{', '|]': '[|', '|}': '{|'}
state = {}

def scan(text_stream):
    global stream, line, char, delims
    line, char = 1, 1
    stream = peekable(text_stream)
    delims = []
    while stream:
        c = stream[0]
        if ord(c) in state:
            scanner = state[ord(c)]
            if scanner is None: continue
            token = scanner(c)
            if token is None: continue
            yield token
        else:
            yield eatwhile(is_name_tail, tag=TokenTag.sym)
    if delims:
        raise Error(delims[-1], f'Unclosed delimiter.')

def add_state(matcher, scanner):
    for match in matcher:
        if len(match) == 1:
            state[ord(match)] = scanner
        elif len(match) == 2:
            for i in range(ord(match[0]), ord(match[1])+1):
                state[i] = scanner

def get():
    global line, char
    x = next(stream)
    if x == '\n':
        line += 1
        char = 1
    else:
        char += 1
    return x

def eatone(tag=-1):
    start = line, char
    if not stream:
        raise Exception('parser error: unexpected EOF') 
    value = get()
    return mktoken(value=value, start=start, stop=(line, char), tag=tag)

def eatwhile(predicate, tag=-1):
    start = line, char
    value = get()
    while stream and predicate(stream[0]):
        value += get()
    return mktoken(value=value, start=start, stop=(line, char), tag=tag)

def eatrange(until, tag=-1, escape=None, literal=False):
    start = line, char
    value = ''
    get()
    while True:
        if not stream:
            tok = mktoken(tag=TokenTag.sym, value='`', start=start, stop=(start[0], start[1] + 1))
            raise Error(tok, f'Unclosed symbol: expected `{chr}`, but found EOF.')
        next = stream[0]
        if next == until: break
        if next == escape:
            get()
            c = get()
            if c == 'n': value += '\n'
            elif c == 't': value += '\t'
            else: value += c
        else:
            value += get()
    get()
    return mktoken(tag=tag, value=value, start=start, stop=(line, char), literal=literal)

def scan_group_open(c):
    start = line, char
    value = get()
    if stream[0] == '|':
        value += get()
    tok = mktoken(value=value, start=start, stop=(line, char), tag=TokenTag.sym)
    delims.append(tok)
    return tok

def check_close_paren(paren: str):
    if not delims:
        raise Error(tok, f'Found closing paren `{paren}`, but there was no open paren.')
    if delims[-1].value != parens[paren]:
        raise Error(tok, f'Incompatible closing paren `{paren}`. Currently open paren is `{delims[-1].value}`.')

def scan_group_close(c):
    tok = eatone(tag=TokenTag.sym)
    check_close_paren(c)
    delims.pop()
    return tok

def scan_group_close_or_symbol(c):
    start = line, char
    value = get()
    if stream[0] in ')]}':
        value += get()
        check_close_paren(value)
        delims.pop()
    else:
        while stream and is_symb_tail(stream[0]):
            value += get()
    return mktoken(value=value, start=start, stop=(line, char), tag=TokenTag.sym)

def scan_newline(c):
    start = line, char
    value = get()
    while stream and stream[0] in '\r\n':
        value += get()
    return mktoken(tag=TokenTag.nl, value=value, start=start, stop=(line, char))

def is_numb_tail(x):
    return '0' <= x <= '9' or x == '_'

def is_name_tail(x):
    return 'a' <= x <= 'z' or 'A' <= x <= 'Z' or '0' <= x <= '9' or x == '_' or x == '-' or ord(x) > 127

def is_symb_tail(x):
    return x in '@%$&=*+!?^/|><:.~-'

add_state('\r'  , None)
add_state('\n'  , scan_newline)
add_state('\t ' , lambda c: eatwhile(lambda x: x in '\t ', tag=TokenTag.ws))
add_state('\\'  , lambda c: eatone(tag=TokenTag.esc))
add_state('#'   , lambda c: eatwhile(lambda x: x != '\n', tag=TokenTag.com))
add_state(',;'  , lambda c: eatone(tag=TokenTag.sym))
add_state('\'\"', lambda c: eatrange(c, tag=TokenTag.str, escape='\\'))
add_state('([{' , scan_group_open)
add_state(')]}' , scan_group_close)
add_state('|'   , scan_group_close_or_symbol)
add_state('`'   , lambda c: eatrange('`', tag=TokenTag.sym, literal=True))
add_state(['09'], lambda c: eatwhile(is_numb_tail, tag=TokenTag.num))
add_state('%$&=*+!?^/:><.~-', lambda c: eatwhile(is_symb_tail, tag=TokenTag.sym))

File: zero/src/scan.nim
import strformat, strutils
import tables

import types

const open_comp = {")": "(", "]": "[", "}": "{", "|)": "(|", "|]": "[|", "|}": "{|"}.to_table
const name_head = {'a'..'z', 'A'..'Z', '_'}
const name_tail = name_head + {'0'..'9', '-'}
const numb_head = {'0'..'9'}
const numb_tail = numb_head + {'a'..'z', 'A'..'Z', '_'}
const symb_head = {'%','$','&','=','*','+','!','?','^','/','>','<',':','.','~','-'}
const symb_tail = symb_head + {'@','|'}

func tos(c: char): string =
    var s = ""
    s.add(c)
    s

type Scanner = object
    x_last, y_last, x, y: int
    i: int
    tmp: string
    text: string
    file: ref string

func new_scanner(str: string, file: ref string): Scanner =
    Scanner(x_last: 1, y_last: 1, x: 1, y: 1, i: 0, tmp: "", text: str, file: file)

proc save(self: var Scanner) =
    self.x_last = self.x
    self.y_last = self.y
    self.tmp = ""

proc eat(self: var Scanner): char =
    let c = self.text[self.i]
    if c == '\n':
        self.y += 1
        self.x = 1
    else:
        self.x += 1
    self.i += 1
    c

func eof(self: Scanner): bool =
    self.i >= self.text.len

func top(self: Scanner, offset: int = 0): char =
    let i = self.i + offset
    if i < self.text.len: self.text[i] else: '\0'

func top_unsafe(self: Scanner, offset: int = 0): char =
    self.text[self.i + offset]

func peek(self: Scanner, offset: int = 1): string =
    self.text[self.i .. self.i + offset]

func get_range(self: Scanner): Position =
    pos(self.y_last, self.x_last, self.y, self.x, self.file)

proc scan*(text: string): seq[Atom] =
    var atoms: seq[Atom]
    var s = new_scanner(text, nil)
    var buf = "" 
    var parens: seq[Atom]
    proc check_close_paren(paren: string) =
        if parens.len == 0:
            echo fmt"Found closing paren `{paren}` but there was no open paren."
            quit(1)
        if parens[^1].value != open_comp[paren]:
            echo fmt"Incompatible closing paren `{paren}`. Currently open paren is `{parens[^1]}`"
            quit(1)

    while not s.eof:
        buf = ""
        s.save
        let curr = s.eat
        buf.add(curr)
        case curr
        of name_head:
            while s.top in name_tail or ord(s.top) > 127: buf.add(s.eat)
            atoms.add(atom(buf, aSym, s.get_range))
        of symb_head:
            while s.top in symb_tail:
                if s.top == '-' and s.top(1) in numb_head:
                    break
                buf.add(s.eat)
            atoms.add(atom(buf, aSym, s.get_range))
        of numb_head:
            while s.top in numb_tail: buf.add(s.eat)
            atoms.add(atom(buf, aNum, s.get_range))
        of {'\'', '"', '`'}:
            buf = ""
            while not s.eof and s.top_unsafe != curr:
                if s.top == '\\':
                    buf.add(s.eat)
                    if s.eof:
                        echo "Unexpected EOF"
                        quit(1)
                buf.add(s.eat)
            if s.eof:
                echo "Unclosed string"
                quit(1)
            discard s.eat
            atoms.add(atom(buf, aStr, s.get_range))
        of {',',';'}:
            atoms.add(atom(buf, aSym, s.get_range))
        of {')',']','}'}:
            check_close_paren(buf)
            discard parens.pop()
            atoms.add(atom(buf, aSym, s.get_range))
        of {'(','[','{'}:
            if s.top == '|': buf.add(s.eat)
            let a = atom(buf, aSym, s.get_range)
            parens.add(a)
            atoms.add(a)
        of '#':
            while s.top != '\n': buf.add(s.eat)
            # TODO: do not discard comments
            #atoms.add(atom(buf, aCom, s.get_range))
        of '|':
            if s.top in {')',']','}'}:
                buf.add(s.eat)
                check_close_paren(buf)
                discard parens.pop()
            else:
                while s.top in symb_tail: buf.add(s.eat)
            atoms.add(atom(buf, aSym, s.get_range))
        of '\\':
            buf.add(s.eat)
            atoms.add(atom(buf, aEsc, s.get_range))
        of {'\t', ' '}:
            while s.top in {'\t', ' '}: buf.add(s.eat)
            atoms.add(atom(buf, aWs, s.get_range))
        of '\n':
            while s.top in {'\r', '\n'}: buf.add(s.eat)
            atoms.add(atom(buf, aNl, s.get_range))
        of '\r':
            discard curr
        else:
            discard curr

    if parens.len != 0:
        echo fmt"Unclosed delimiter `{parens[^1]}`"
        quit(1)

    atoms

proc indent_error(levels: seq[Atom], next: Atom) =
    echo "Indentation error"
    quit(1)

proc indent*(tokens: seq[Atom]): seq[Atom] =
    var res: seq[Atom]
    var levels: seq[Atom]
    var i = 0

    proc pop_levels(n: int) =
        var tok: Atom
        for _ in 0..<n:
            tok = levels.pop()
            res.add(Atom(value: "$DED", tag: aDed, pos: tok.pos))
        if i < tokens.len:
            tok = tokens[i]
            res.add(Atom(value: "$CNT", tag: aCnt, pos: tok.pos))

    proc find_level(match: string): int =
        for j in countdown(high(levels), 0):
            if levels[j].value == match:
                return j
        -1

    while i < tokens.len:
        let t = tokens[i]
        i += 1
        if t.tag notin {aWs, aNl}:
            res.add(t)

        elif t.tag == aNl and i < tokens.len:
            let next = tokens[i]
            let next_level = next.value
            let last_level = if levels.len > 0: levels[^1].value else: ""
            if next.tag != aWs:
                pop_levels(levels.len)
                res.add(next)
            elif i < tokens.len and tokens[i].tag == aNl:
                continue
            elif next_level == last_level:
                res.add(Atom(value: "$CNT", tag: aCnt, pos: next.pos))
            elif next_level.starts_with(last_level):
                levels.add(next)
                res.add(Atom(value: "$IND", tag: aInd, pos: next.pos))
            elif last_level.starts_with(next_level):
                let level = find_level(next_level)
                if level == -1: indent_error(levels, next)
                pop_levels(level)
            else:
                indent_error(levels, next)

    pop_levels(levels.len)
    res